src/main/java/org/cirdles/topsoil/app/TSVTable.java

```diff
@@ -42,15 +55,96 @@
 
     private Path savePath;
     private Dataset dataset;
+    private ObservableList<TableRow<Entry>> deselectedRows;
 
     public TSVTable() {
         this.setEditable(true);
+        this.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);
         
         this.setOnKeyPressed((KeyEvent event) -> {
             if (event.isShortcutDown() && event.getCode().equals(KeyCode.V)) {
                 pasteFromClipboard();
             }
         });
+        
+        List<TableRow<Entry>> rows = new ArrayList<>(); //List of selected rows
+        deselectedRows = FXCollections.<TableRow<Entry>>observableArrayList();
+               
+        //Right-Click Menu
+        this.setRowFactory((TableView<Entry> tableView) -> {
+            final TableRow<Entry> row = new TableRow<>();   
+            rows.add(row);    //Multiple row selection  
+            
+            final ContextMenu rowMenu = new ContextMenu();
+            MenuItem deselectItem = new MenuItem("Deselect");
+            deselectItem.setOnAction((ActionEvent event) -> {
+                tableView.getSelectionModel().getSelectedIndices().stream() //For each row selected
+                        .map(index -> index + 1) //indexes are offsetted
+                        .map(rows::get)
+                        .forEach(tableRow -> {
+                            if(tableView.getSelectionModel().getSelectedItem().getIsSelected()) { //if IsSelected is true                             
+                                deselectedRows.add(tableRow);
+                            } else {
+                                deselectedRows.remove(tableRow);
+                            }
+                        });      
+                this.getSelectionModel().clearSelection();
+            });                
+
+            //Add context menu
+            rowMenu.getItems().addAll(deselectItem);
+            
+            // only display context menu for non-empty rows
+            row.contextMenuProperty().bind(
+                    Bindings.when(Bindings.isNotNull(row.itemProperty()))
+                            .then(rowMenu)
+                            .otherwise((ContextMenu) null));
+                                 
+            return row;
+        });        
```

Comment:
You can't have a lambda function this big. Please make a class for this and break the logic up into methods.